### PR TITLE
[SPARK-13528][SQL] Make the short names of compression codecs consistent in ParquetRelation

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

This pr to make the short names of compression codecs in `ParquetRelation` consistent against other ones. This pr comes from #11324.
## How was this patch tested?

Add more tests in `TextSuite`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -60,6 +60,51 @@ private[spark] object CallSite {
   val empty = CallSite("", "")
 }
 
+/** An utility class to map short compression codec names to qualified ones. */
+private[spark] class ShortCompressionCodecNameMapper {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: I agree with standardizing names, but this seems like a lot of over-engineering, with abstract classes and whatnot just to attempt to make some keys consistent. I think this makes it harder to understand, and would stick to standardizing the keys.

Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Yeah, in an original idea, we should have consistent short names for compression codecs everywhere  in spark. Any simpler way you have?

**Comment by maropu** [comment]
@rxin ping

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 1 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: +1, agree with Sean.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: I also agree - this abstract class is too much. I think just having lz4/bzip2 etc in different places isn't that big of a deal.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala [hunk 2]
```diff
@@ -58,18 +58,21 @@ class TextSuite extends QueryTest with SharedSQLContext {
...
 
-    val tempFile = Utils.createTempDir()
-    tempFile.delete()
-    df.write
-      .option("compression", "gZiP")
-      .text(tempFile.getCanonicalPath)
-    val compressedFiles = tempFile.listFiles()
-    assert(compressedFiles.exists(_.getName.endsWith(".gz")))
-    verifyFrame(sqlContext.read.text(tempFile.getCanonicalPath))
+    Seq("bzip2", "deflate", "gzip").map { codecName =>
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: Nit: `foreach` not `map`. It doesn't matter in practice though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 3]
```diff
@@ -905,9 +905,9 @@ private[sql] object ParquetRelation extends Logging {
 
   // The parquet compression short names
   val shortParquetCompressionCodecNames = Map(
-    "NONE" -> CompressionCodecName.UNCOMPRESSED,
-    "UNCOMPRESSED" -> CompressionCodecName.UNCOMPRESSED,
-    "SNAPPY" -> CompressionCodecName.SNAPPY,
-    "GZIP" -> CompressionCodecName.GZIP,
-    "LZO" -> CompressionCodecName.LZO)
+    "none" -> CompressionCodecName.UNCOMPRESSED,
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Does this cause any compatibility problems? I don't know this code well. If these values are never used outside Spark itself then it's fine of course.

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: ISTM there are few people who use this option, what do u think? cc: @rxin

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Did this change any behavior?

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Ah, ... no. This pr does not change it.

**Comment by rxin** [comment]
@HyukjinKwon can you help review this one? It looks ok to me. Maybe you can do yours on top of this, which adds Python and better error messages.

**Comment by HyukjinKwon** [comment]
@rxin Sure.

**Comment by HyukjinKwon** [comment]
One thing I am not sure here is, `DeflateCodec` is basically zlib as far as I remember but `.deflate` is Hadoop's conversion. For example, for ORC, it uses `ZLIB` rather than `Deflate`. Although I agree it is a good idea to make the names consistent, I am a bit worried if they are safe or not.

Can we maybe better make them `private`?

**Comment by HyukjinKwon** [comment]
One more thing is, I am not sure if we then need `uncompressed` or `none` for JSON, CSV and TEXT datasources as shorten names to explicitly set no compression codec.

Would you give me some feedback please?

**Comment by HyukjinKwon** [comment]
If the consistent short names infer only lower-case names, then I think we can just leave them as corrected here in a way.

**Comment by maropu** [comment]
Yes, you're right . `gzip` and `deflate` has a compression algorithm in common. However, in my opinion, it'd be better to support both short names for hadoop compatibility because the file formats for them are different.

**Comment by maropu** [comment]
Setting no compression codec explicitly is important for uses to understand behaviours though, I'm not sure we need both names...

**Comment by rxin** [comment]
I think "none" is enough.

**Comment by HyukjinKwon** [comment]
Thanks. Then `zlib` for ORC, `deflate` for Hadoop and `none` for TEXT, JSON, CSV and ORC?

**Comment by rxin** [comment]
Sorry I don't understand what you mean ...

**Comment by HyukjinKwon** [comment]
Let me list up possible codec optionsn for each data source.

For JSON, CSV and TEXT data sources,
`none` - no compression
`bzip2`
`snappy`
`gzip`
`deflate` - this is zlib

For Parquet,
`none`/`uncompressed` - 'uncompressed' was possible to set to Spark configuration option.
`gzip`
`lzo`
`snappy`
`deflate`

For ORC,
`none` - no compression 
`zlib`
`snappy`
`lzo`

**Comment by rxin** [comment]
Why do we need to support uncompressed in parquet? It it for backward compatibility?

**Comment by HyukjinKwon** [comment]
Yes would the users set 'uncompressed' as it was possible to set via Spark configuration?

**Comment by rxin** [comment]
OK sounds great. I'd have it as an undocumented way for backward compatibility.

**Comment by rxin** [comment]
deflate is pretty confusing. I'd just say zlib?

**Comment by HyukjinKwon** [comment]
As I said then we might beed to consider handling the extensions for JSON, TEXT and CSV which are `.deflate`. We might need to manually change them to `.zlib

**Comment by rxin** [comment]
We don't specify any extensions right now, do we?

**Comment by HyukjinKwon** [comment]
I remember I saw some tests for compression codes which check the extension when they are compressed. Then, let me correct them (or simply check them)and then create a new PR based on this.

**Comment by HyukjinKwon** [comment]
Let's talk more in the new PR. I will try to deal with this in capacity myself first.

**Comment by rxin** [comment]
OK I thought about this a little bit more -- I'd just have uncompressed as an undocumented option for all data sources. That way, it is very consistent.

@HyukjinKwon should I merge this pr now?

**Comment by HyukjinKwon** [comment]
Yes please. Let me make a followup.

**Comment by rxin** [comment]
Thanks - merging this in master.

**Pull request closed without merging**